docs/decisions/0007-lti-1p3-decoupling.rst

```diff
@@ -0,0 +1,115 @@
+the only context in which LTI launches occurred.
+
+The dependence on the ``location`` field prevents LTI 1.3 launches from being possible outside the context of an XBlock.
+This prevents Open edX from using LTI more widely across the platform and across the ecosystem, because its use is
+restricted to specific contexts in which ``location`` is meaningful.
+
+For example, because ``LTIConfiguration`` instances are associated with a particular ``location``, it is not possible for a
+``LtiConfiguration`` to be reused across different XBlocks (i.e. a course context), even if the ``LtiConfiguration`` would
+be identical amongst the different XBlocks.
+
+In an LTI 1.3 launch, the ``location`` field is used to determine the ``https://purl.imsglobal.org/spec/lti/claim/roles``
+and ``https://purl.imsglobal.org/spec/lti/claim/context`` claims of the LTI 1.3 launch. In an LTI 1.3 launch, the
+``https://purl.imsglobal.org/spec/lti/claim/roles`` claim is required, and the
+``https://purl.imsglobal.org/spec/lti/claim/context`` claim is optional. This couples the LTI 1.3 launches to the XBlock
+through the ``location`` field.
+
+Furthermore, LTI 1.3 launches are also dependent on functions imported from edx-platform. These functions are used `to
+get the requesting user's role
+<https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/plugin/views.py#L182>`_, `to get the requesting
+user's external ID <https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/plugin/views.py#L183>`_, `to
+get the course key from the location field
+<https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/plugin/views.py#L181>`_, etc. This prevents LTI
+1.3 launches from running outside of the platform.
+
+Open edX would like to support the following LTI use cases:
+
+* content launches in an XBlock context
+* content launches from multiple XBlocks, reusing LTIConfigurations between them (i.e. course context)
+* content launches outside an XBlock context in the platform (e.g. LTI course tabs, LTI 1.3 embedded discussions)
+* content launches from an independently deployable application (i.e. IDA or Django context)
+
+https://github.com/openedx/xblock-lti-consumer/pull/254 decoupled the LTI 1.3 launch handler endpoint
+``launch_gate_endpoint``, the LTI 1.3 keyset endpoint ``public_keyset_endpoint``, and the LTI 1.3 access token endpoint
+``access_token_endpoint`` from the XBlock by moving these endpoints into the Django plugin. This was the first step in
+decoupling the XBlock from the LTI 1.3 launch. This decision describes how to fully decouple the LTI 1.3 launch from the
+XBlock.
+
+Decision
+********
+
+* We will remove references to the ``location`` field from all views that implement the generic LTI 1.3 launch. This
+  includes the ``launch_gate_endpoint``, ``public_keyset_endpoint``, and ``access_token_endpoint`` views.
```

Comment:
@MichaelRoytman @Zacharis278 The `access_token_endpoint` needs contextual information in order to select which scopes and services it'll allow the tool to use. [Check the access_token method here](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p3/consumer.py#L343).

For this, we should take the simple approach and store the LTI Advantage service configuration directly in the LTIConfiguration model and let all instances re-using that configuration have access to the same set of services (instead of storing LTI Adv configuration for each placement).

What do you say?

Reply:
@giovannicimolin 

> The access_token_endpoint needs contextual information in order to select which scopes and services it'll allow the tool to use. [Check the access_token method here](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p3/consumer.py#L343).

What additional contextual information does this endpoint need? I see that it needs a `UsageKey` or a `config_id` to fetch the corresponding `LtiConfiguration` instance, but is there any additional context that's not already stored in the `LtiConfiguration`? I'm not seeing other necessary information. Can you point me at it? It looks like the library grants all scopes that the consumer/tool requests [here](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p3/consumer.py#L386), without referring to any contextual information. I am not suggesting that we remove the use of the `LtiConfiguration` model. Rather, I'm suggesting that we refer to `LtiConfigurations` by their `config_id` instead of their `location`.

> For this, we should take the simple approach and store the LTI Advantage service configuration directly in the LTIConfiguration model and let all instances re-using that configuration have access to the same set of services (instead of storing LTI Adv configuration for each placement).

This is referring to the case where multiple placements of an LTI link will share a single `LtiConfiguration`, right?

If multiple LTI link placements share a single `LtiConfiguration`, so long as `get_lms_lti_access_token_link(location)` is called with the same `config_id` when generating the access token to provide to the tool, the same `LtiConfiguration` instance will be used to generate the access token for the tool, so that should address what you're saying.

Eventually, we'd like for `get_lms_lti_access_token_link` to only accept a `config_id`.


I am not very familiar with LTI Advantage services just yet, so please let me know if I'm misunderstanding something.

Reply:
@MichaelRoytman 

>What additional contextual information does this endpoint need? I see that it needs a UsageKey or a config_id to fetch the corresponding LtiConfiguration instance, but is there any additional context that's not already stored in the LtiConfiguration? I'm not seeing other necessary information. Can you point me at it? 

You're right! It currently grants all available scopes, but there's a `TODO` in the code [here](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/lti_1p3/consumer.py#L424) that mentions limiting scopes.

>This is referring to the case where multiple placements of an LTI link will share a single LtiConfiguration, right?

Yes, and you are right about this as well. The LTI configuration should handle the advantage configuration, while the tool just needs to use the right `ltiResourceLink` to get and set grades for each placement.

Reply:
My original concerns were not with the token, but with the access levels that tools will have to learner data (this is the question I asked myself: does a tool configured to work platform-wide have access to **all users** through NRPS? I don't think it should). 

But I think this will be a different problem to tackle in a separate ADR. Lets keep this one about decoupling LTI 1.3 and handle the permission issues separately. Right now our main use case (XBlock) won't cause issues.